devito/types/petsc.py

```diff
@@ -57,7 +59,16 @@ class KSPConvergedReason(LocalObject):
     dtype = CustomDtype('KSPConvergedReason')
 
 
-class PETScArray(ArrayBasic):
+class PETScFunction(Differentiable):
+    """
+    Base class for all PETSc objects.
+    Differentiable enables compatability with standard Function objects,
+    allowing for the use of the `subs` method.
+    """
+    pass
+
+
+class PETScArray(ArrayBasic, PETScFunction):
```

Comment:
So why not just `DiscreteFunction`  which is basically `Differentiable+AbstractFunction` ?

Reply:
The PETSc objects need to be treated more like arrays, in the sense that they are compiler generated and do not necessarily carry user data.